### ISSUE TITLE
Seeking recommendations for elegantly showing end of results / no results messages

### ISSUE THREAD

**Issue by dpraul** [dhilt/ngx-ui-scroll#273]
Hi there! Thanks again for your work on this library - we've been using it for some time now and have been very pleased.

My only complaint is everything I do to try to expose "end of results" or "no results" messages feels very hacky. Every one of my use-cases for this library is with finite remote datasources, so these are absolutely necessary to show users that there is no more scrolling or nothing to see.

Ideally, I'd like these messages to be shown within the scroll viewport alongside the `uiScroll` items, but every way I've attempted to do so has run into issues, mostly because of the `[data-padding-backward]` and `[data-padding-forward]`. For example, showing "no more items" directly after the last item leaves a large empty gap if there are fewer items than fits the viewport, because the `[data-padding-forward]` sets its height to fill the rest of the viewport. Similarly, adding a "No results found" into the viewport causes a scroll area to appear for the same reason as the prior example.

The workarounds I've done to display these messages have been to take them out of the viewport and:
- Position the "no results" message absolutely over the scroll area, and only show it when `adapter.isLoading$` emits `false` and `adapter.itemsCount === 0`
- Position the "end of results" message absolutely at the bottom of the scroll area, and only show it when `adapter.eof$` emits `true`, `adapter.itemCount > 0`, and the user scrolls to the bottom.

Neither of these are especially elegant nor simple to implement, and are prone to issues.

Long story short, is there any way you could recommend to implement these messages more elegantly? Or is this a limitation of the library, and I'm simply trying to make it do things that it wasn't built for?

**Comment by dhilt** [comment]
@dpraul A scrollable viewport is a space for items (real or virtual) to be scrolled. From this point of view I would suggest the following approaches.

**1.** No results. If you know that there are no items to scroll at the component level, you may provide two separate views that may be similar to each other.

```html
<div class="viewport" *ngIf="empty">
  <div class="empty"> No results </div>
</div>
<div class="viewport" *ngIf="!empty">
  <div *uiScroll="...">  ...  </div>
</div>
```

Another option is to implement the datasource with only one special item.

```
constructor () {
  this.items = [];
  this.items.push({ empty: true });
}

datasource = new Datasource({
  get: (start, count, done) => {
    let data = [];
    const end = Math.min(start + count - 1, this.items.length - 1);
    if (start <= end) {
      data = this.items.slice(start, end + 1);
    }
    done(data);
  }
})
```

```html
<div class="viewport">
  <div *uiScroll="let item of datasource">
    <div *ngIf="item.empty" class="empty"> No results </div>
    <div *ngIf="!item.empty">{{item.text}}</div>
  </div>
</div>
```

**2.** End of results. The EOF massage (end of file) could be treated as some special item or special set of items.

```ts
constructor () {
  this.items = Array.from({ length: 1000 }).map((i, j) => ({
    text: 'item #' + (index + j)
  }));
  this.items.push({ last: true });
}

datasource = new Datasource({
  get: (start, count, done) => {
    let data = [];
    const end = Math.min(start + count - 1, this.items.length - 1);
    if (start <= end) {
      data = this.items.slice(start, end + 1);
    }
    done(data);
  }
})
```

```html
<div class="viewport">
  <div *uiScroll="let item of datasource">
    <div *ngIf="item.last" class="last"> End of results </div>
    <div *ngIf="!item.last">{{item.text}}</div>
  </div>
</div>
```

So the Scroller's HTML contains only items, but these items may have different quality: common item, "eof" item, "no items" item...

**Comment by dpraul** [comment]
Thanks for the response! I really appreciate the detailed notes and suggestions. I tried the approach with two separate viewports but ran into errors with getting the data stream _back_ onto the page after changing inputs and resetting (it looks like there are commits in the most recent version that might fix that, but we're on Angular 11 for the time being). 

That said, mixing the "eof" and "no items" messages into the data stream seems like it would still work. I've been playing around with your suggestions and have almost got something working, it's just a bit tricky with a remote data source (since the logic to find the end of the source relies on knowing the full length of the results). The following is close:

```ts
type RenderItem = { content: text; } | { end: true; };

class Component {
  datasource = new Datasource<RenderItem>({
    get: (offset: number, count: number) => this.queryItems(count, offset).pipe(
      map((items) => {
        if (items.length === 0) {
          return [{ end: true }];
        }
        if (items.length < count) {
          return [...items, { end: true };
        }
        return items;
      });
    ),
  });
}
```

and then I formed the template something like this:
```html
<div class="viewport">
  <div *uiScroll="let item of datasource">
    <div *ngIf="item.content">{{ item.content }}</div>
    <div *ngIf="item.end && datasource.adapter.itemsCount > 1">No more items!</div>
    <div *ngIf="item.end && datasource.adapter.itemsCount === 1">No results found</div>
  </div>
</div>
```

This works in most cases, except when the remote call returns exactly `count - 1` items. That causes the `{ end: true }` to get appended, meaning the `Datasource` gets the count it expected, and when it makes the next call it gets another `{ end: true }` since the remote data has no more items, causing two "No more items!" messages. I suppose I could store some state outside the Datasource for whether the end message has already been returned, that just feels like yet another hack.

**Comment by dhilt** [comment]
@dpraul First, Angular 12 updates relate to only demo app, and ngx-ui-scroll should work with Angular 11 as well as it works with Angular 12. You are right, the last release is dedicated to clear resetting of the Scroller and ngIf is the most common case for that fix.

Second, your code looks nice, mapping the remote result the way you did it is how I would do it by myself. The fix for the edge case you mentioned can be resolved by setting `maxIndex`. 

Opt.1. If you know the length of the remote dataset, you may just provide appropriate value to the Datasource settings object on initialization:

```js
  datasource = new Datasource<RenderItem>({
    get: ...,
    settings: {
      startIndex: 1,
      minIndex: 1,
      maxIndex: REMOTE_COUNT + 1 // + 1 for "eof" item
    }
  });
```

Opt.2. Remote data counter is unknown. You may use Adapter.fix method, which is a part of [experimental API](https://dhilt.github.io/ngx-ui-scroll/#/experimental), and set `maxIndex` in runtime:

```ts
  datasource = new Datasource<RenderItem>({
    get: (offset: number, count: number) => this.queryItems(count, offset).pipe(
      map((items) => {
        // I believe this branch is not necessary, the next one covers it
        // if (items.length === 0) {
        //   return [{ end: true }];
        // }
        if (items.length < count) {
          this.datasource.adapter.fix({ maxIndex: offset + items.length + 1 });
          return [...items, { end: true };
        }
        return items;
      });
    ),
  });
```

This last option is very interesting, I would like to have a spec for this case when publishing `Adapter.fix({ minIndex })`... Hope it works this way, I would expect this. If not, we'll see what can be done.

**Comment by dpraul** [comment]
`Adapter.fix` looks like exactly what I was missing, and the code you provided works almost perfectly - thank you so much. I made two changes: to use `switchMap` so that the response waits for `fix` to complete, and took out the `+ 1` since `length` starts at 1 but `maxIndex` starts at 0.

```ts
type RenderItem = { content: text; } | { end: true; };

class Component {
  datasource = new Datasource<RenderItem>({
    get: (offset: number, count: number) => this.queryItems(count, offset).pipe(
      switchMap(async (items) => {
        if (items.length < count) {
          await this.datasource.adapter.fix({ maxIndex: offset + items.length });
          return [...items, { end: true }];
        }
        return items;
      })
    ),
  });
}
```

---
EDIT:
For anyone else using an `Observable` get, I was able to generalize this as a pretty simple-to-use RxJS operator:

```ts
import { OperatorFunction } from 'rxjs';
import { switchMap } from 'rxjs/operators';

import type { IDatasource } from 'ngx-ui-scroll';

export const UI_SCROLL_END_MESSAGE = { end: true } as const;

/**
 * Used with ngx-ui-scroll to add an "end of results" message to the Datasource response as needed.
 *
 * @example Component
 *   class ExampleComponent {
 *     datasource = new Datasource({
 *       get: (offset: number, count: number) => this.remoteGet$(count, offset).pipe(
 *         addUiScrollEndMessage(this.datasource, count, offset)
 *       ),
 *     });
 *   }
 *
 *
 * @example Template
 *   <div class="ui-scroll-viewport">
 *     <div *uiScroll="let item of datasource" class="scroll-item">
 *       <div *ngIf="!item.end">{{ item.content }}</div>
 *       <div *ngIf="item.end && notesDatasource.adapter.itemsCount > 1">End of results</div>
 *       <div *ngIf="item.end && notesDatasource.adapter.itemsCount === 1">No results</div>
 *     </div>
 *   </div>
 */
export function addUiScrollEndMessage<T extends any[]>(
  datasource: IDatasource,
  count: number,
  offset: number
): OperatorFunction<T, T | typeof UI_SCROLL_END_MESSAGE[]> {
  return (source$) => source$.pipe(
    switchMap(async (items) => {
      if (items.length < count) {
        await datasource.adapter.fix({ maxIndex: offset + items.length });
        return [...items, UI_SCROLL_END_MESSAGE];
      }
      return items;
    })
  );
}
```

**Comment by dhilt** [comment]
@dpraul This looks very nice and very helpful! One thing slips away from me, the thing is the logic of the `Adapter.fix` method is synchronous. Its promise resolves immediately, and there should be no need in awaiting it.

```ts
console.log('Start');
datasource.adapter.fix({ maxIndex: 100 }).then(result => console.log('result', result));
console.log('Done');
```

Output with `devSettings.debug` enabled:

```
Start
adapter: fix({ maxIndex })
fire adapter.fix "start" {options: {…}}
run adapter.fix {maxIndex: 100}
fire adapter.fix "done"
Done
result {success: true, immediate: true, details: null}
```

Of course, the "result" output goes last because of promise wrapping, but setting `maxIndex` runs and logs synchronously. Maybe it can be simplified? There may be another potential problems with `Adapter.fix({ maxIndex })` if this `await` is really needed. Also, I wonder if it is possible to create a short demo on stackblitz implementing this approach (by forking [this demo](https://stackblitz.com/edit/ngx-ui-scroll-append-async) for example)...

**Comment by dpraul** [comment]
It should be no issue whether the `adapter.fix` API is synchronous or asynchronous! I have the operator that way to maintain type compatibility, so it should work regardless.

I went ahead and created the demo you requested :) It is available here: https://stackblitz.com/edit/ngx-ui-scroll-end-of-results-demo

**Comment by dhilt** [comment]
@dpraul The issue is still open and there is a good reason:

![ngx-ui-scroll-end-of-results-demo-forked](https://user-images.githubusercontent.com/4365660/127146082-b406a49c-1866-4790-b7e0-628b10569f1f.gif)

This edge case should be fixed before... and it should take some time

**Comment by dpraul** [comment]
@dhilt good catch! Looks like there was a race condition that allowed the `maxIndex` to be set multiple times. I've added a check to make sure it only gets set once, and I can no longer reproduce the issue you showed. Let me know if you find anything else!

**Comment by dhilt** [comment]
Well, this is done, the final demo is here: **https://stackblitz.com/edit/ngx-ui-scroll-end-of-results-demo-forked**.
